library/ssl_tls.c

```diff
@@ -2822,20 +2822,49 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
 
 /*
  * Retransmit the current flight of messages.
+ */
+int mbedtls_ssl_resend( mbedtls_ssl_context *ssl )
+{
+    int ret = 0;
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> mbedtls_ssl_resend" ) );
+
+    ret = mbedtls_ssl_flight_transmit( ssl );
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= mbedtls_ssl_resend" ) );
+
+    return( ret );
+}
+
+/*
+ * Transmit or retransmit the current flight of messages.
  *
  * Need to remember the current message in case flush_output returns
  * WANT_WRITE, causing us to exit this function and come back later.
  * This function must be called until state is no longer SENDING.
  */
-int mbedtls_ssl_resend( mbedtls_ssl_context *ssl )
+int mbedtls_ssl_flight_transmit( mbedtls_ssl_context *ssl )
 {
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> mbedtls_ssl_resend" ) );
+    const int ret_payload = mbedtls_ssl_get_max_out_record_payload( ssl );
```

Comment:
Blocker: This value depends on the activated outgoing transform, which changes during flight transmission. While during the initial handshake, this rather leads to an underapproximation of the available space for the payload, during a renegotiation it might be a strict overapproximation, leading to packets of size larger than the configured MTU to be emitted. I tested this with a connection initially using AES and then switching to ChaChaPoly, and using long session ticket that needs fragmentation. In this case, the available space for the SessionTicket is judged by the maximum payload size in the ChaChaPoly suite, but it's actually encrypted using the AES transform.

This can be fixed by re-calculating `ret_payload` and `max_hs_fragment_len` in each iteration, after a potential swapping of transforms.

Reply:
Nice catch! Fixing.